crates/bevy_math/src/sampling/mesh_sampling.rs

```diff
@@ -0,0 +1,68 @@
+//! Functionality related to random sampling from triangle meshes.
+
+use crate::{primitives::Measured2d, primitives::Triangle3d, ShapeSample, Vec3};
+use rand::Rng;
+use rand_distr::{Distribution, WeightedAliasIndex};
+use thiserror::Error;
+
+/// A [distribution] that caches data to allow fast sampling from a collection of triangles.
+/// Generally used through [`sample`] or [`sample_iter`].
+///
+/// [distribution]: Distribution
+/// [`sample`]: Distribution::sample
+/// [`sample_iter`]: Distribution::sample_iter
+///
+/// Example
+/// ```
+/// # use bevy_math::{Vec3, primitives::*};
+/// # use bevy_math::sampling::mesh_sampling::UniformMeshSampler;
+/// # use rand::{SeedableRng, rngs::StdRng, distributions::Distribution};
+/// let faces: Vec<Triangle3d> = Tetrahedron::default().faces().into();
+/// let sampler = UniformMeshSampler::try_new(faces).unwrap();
+/// let rng = StdRng::seed_from_u64(8765309);
+/// // 50 random points on the tetrahedron:
+/// let samples: Vec<Vec3> = sampler.sample_iter(rng).take(50).collect();
+/// ```
+pub struct UniformMeshSampler {
+    triangles: Vec<Triangle3d>,
+    face_distribution: WeightedAliasIndex<f32>,
+}
+
+impl Distribution<Vec3> for UniformMeshSampler {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        let face_index = self.face_distribution.sample(rng);
+        self.triangles[face_index].sample_interior(rng)
+    }
+}
+
+/// An error that indicates that a [`UniformMeshSampler`] could not be constructed
+/// because its input data had a total area of zero.
+#[derive(Debug, Error)]
+#[error("Failed to form distribution: provided triangles have zero area")]
+pub struct ZeroAreaMeshError;
+
+impl TryFrom<Vec<Triangle3d>> for UniformMeshSampler {
```

Comment:
I actually ended up just getting rid of the `TryFrom` trait implementation, but I made the `try_new` function take an `IntoIterator<Item = Triangle3d>` instead of an `Into<Vec<Triangle3d>>`. It's hard to make `TryFrom` more generic (blanket impl in `core` basically prevents any blanket implementation you would ever dream up), and I think its value was already a little dubious to begin with.